cmd/bpf2go/vfs.go

```diff
@@ -0,0 +1,224 @@
+	ExternalContents string      `json:"external-contents,omitempty"`
+
+	// specificity is used internally to resolve conflicts
+	//
+	// We populate vfs from go modules. Surprisingly, module file
+	// trees may overlap, e.g. "github.com/aws/aws-sdk-go-v2" and
+	// "github.com/aws/aws-sdk-go-v2/internal/endpoints/v2".
+	specificity int
+}
+
+type vfsItemType string
+
+const (
+	vfsFile           vfsItemType = "file"
+	vfsDirectory      vfsItemType = "directory"
+	vfsDirectoryRemap vfsItemType = "directory-remap"
+)
+
+// vfsAdd adds "directory-remap" entry for dir under vpath
+func vfsAdd(root *vfsItem, vpath, dir string) error {
+	return vfsAdd2(root, vpath, dir, vfsDirectoryRemap, len(vpath), defaultFS{})
+}
+
+type defaultFS struct{}
+
+func (defaultFS) Open(name string) (fs.File, error) {
+	return os.Open(name)
+}
+
+func (defaultFS) Stat(name string) (fs.FileInfo, error) {
+	return os.Stat(name)
+}
+
+func vfsAdd2(vi *vfsItem, vpath, path string, typ vfsItemType, specificity int, fs fs.FS) error {
+	for _, name := range strings.Split(vpath, "/") {
+		if name == "" {
+			continue
+		}
+		idx := vi.index(name)
+		if idx == -1 {
+			switch vi.Type {
+			case vfsDirectoryRemap:
+				newItem := vfsItem{Name: vi.Name, Type: vfsDirectory}
+				if err := vfsPopulateFromDir(&newItem, vi.ExternalContents, vi.specificity, fs); err != nil {
+					return err
+				}
+				*vi, idx = newItem, newItem.index(name)
+			case vfsFile:
+				vi.Type, vi.ExternalContents = vfsDirectory, ""
+			}
+			if idx == -1 {
+				idx = len(vi.Contents)
+				vi.Contents = append(vi.Contents, vfsItem{Name: name, Type: vfsDirectory})
+			}
+		}
+		vi = &vi.Contents[idx]
+	}
+	switch vi.Type {
+	case vfsDirectory:
+		if len(vi.Contents) == 0 {
+			vi.Type, vi.ExternalContents, vi.specificity = typ, path, specificity
+			return nil
+		}
+		if typ == vfsFile {
+			return nil
+		}
+		return vfsPopulateFromDir(vi, path, specificity, fs)
+	case vfsDirectoryRemap, vfsFile:
+		if vi.specificity <= specificity {
+			vi.Type, vi.ExternalContents, vi.specificity = typ, path, specificity
+		}
+	}
+	return nil
+}
+
+func vfsPopulateFromDir(vi *vfsItem, dir string, specificity int, fsys fs.FS) error {
+	entries, err := fs.ReadDir(fsys, dir)
+	if err != nil {
+		return err
+	}
+	for _, entry := range entries {
+		path := filepath.Join(dir, entry.Name())
+		info, err := fs.Stat(fsys, path) // follows symlinks
+		if err != nil {
+			return err
+		}
+		typ := vfsFile
+		if info.Mode().IsDir() {
+			typ = vfsDirectoryRemap
+		}
+		if err := vfsAdd2(vi, entry.Name(), path, typ, specificity, fsys); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (vi *vfsItem) index(name string) int {
+	return slices.IndexFunc(vi.Contents, func(item vfsItem) bool {
+		return item.Name == name
+	})
+}
+
+// persistVfs stores vfs in user cache dir under a path derived from
+// outputDir; the file should stay around for the benefit of a language
+// server / IDE
```

Comment:
Other people on my team use it. I think there's a tool you can use for launching other programs, `bpf2go` included. The tool does some `LD_PRELOAD` hackery to capture arguments passed to a compiler. This is later used by `clangd` to configure include paths, amongst other things. I will share the tool name shortly.

Basically, it is autocompletion that stops working, plus plenty of bogus compile errors are shown.

Vfsoverlay file has to stay oround, otherwise `clangd` will have a very different idea about include paths than the actual compiler used to build bpf code.

Reply:
https://github.com/rizsotto/Bear